### PR TITLE
Fix slow mask processing for causal masks.

### PR DESCRIPTION
**Pull request by huytransformer**
# Description

Currently _process_mask builds every mask block in a Python loop to classify it as empty/partial/full. This is quadratic in sequence length eg ~13 min for a 1M causal mask

This PR 1) proposes skipping the loop for causal masks and also works for permuted q_sequence (load balanced CP). 
2) makes qk_diag_skip raise for permuted q_sequence, nonzero offset, dynamic masks, and ring attention, where it would silently produce wrong outputs.

### Performance comparison

`process_mask` wall time for a causal mask with a permuted `q_sequence`, block shape (512, 512)

| Sequence length | Before | After |
|---|---|---|
| 64K | 3.0 s | 0.001 s |
| 128K | 14.5 s | 0.001 s |
| 256K | 47.4 s | 0.004 s |
| 1M | 755 s | 0.067 s |

# Test
`pytest tokamax/_src/ops/experimental/tpu/splash_attention/splash_attention_mask_test.py splash_attention_kernel_test.py ring_attention_kernel_test.py`

```
splash_attention_mask_test.py::SplashAttentionMaskInfoTest::test_causal_mask_with_permuted_q_sequence PASSED
splash_attention_mask_test.py::SplashAttentionMaskInfoTest::test_rectangular_wide_causal_mask0 PASSED
splash_attention_mask_test.py::SplashAttentionMaskInfoTest::test_rectangular_tall_causal_mask0 PASSED
splash_attention_mask_test.py::SplashAttentionMaskInfoTest::test_causal_two_q_shards_two_kv_shards0 PASSED
ring_attention_kernel_test.py::RingAttentionValidationTest::test_qk_diag_skip_rejected PASSED
...
25 passed, 456 skipped
```

### CODE REVIEW

File: tokamax/_src/ops/experimental/tpu/splash_attention/splash_attention_mask_info.py [hunk 1]
```diff
@@ -441,31 +477,36 @@ def _process_mask(
...
-      if chunk.all():
-        state_grid[q_idx, kv_idx] = 2
-      else:
-        state_grid[q_idx, kv_idx] = 1
-        chunk_id = partial_blocks_map[_HashableNDArray(chunk)]
-        partial_id_grid[q_idx, kv_idx] = chunk_id
-
-        if chunk_id == len(unique_chunks):
-          unique_chunks.append(chunk)
+  if isinstance(mask, mask_lib.CausalMask):
```
Comment (octatrifan) [inline tokamax/_src/ops/experimental/tpu/splash_attention/splash_attention_mask_info.py:485]: Am I correct to assume that the same block classification logic can be adapted to work for `LocalMask` as well? Outside the scope of this PR, but could be a follow-up as it covers another common case.

File: tokamax/_src/ops/experimental/tpu/splash_attention/splash_attention_kernel.py [hunk 2]
```diff
@@ -2186,16 +2186,26 @@ def _make_splash_attention(
...
-    # The skip assumes kv > q is ALWAYS masked — a pure-causal property. Any mask
-    # that admits a valid kv > q entry (bidirectional, local/sliding window, custom)
-    # would be silently corrupted, so fail loud. (Square-block preconditions are
-    # enforced in SplashConfig.__post_init__.)
-    raise ValueError(
-        "qk_diag_skip=True requires a pure CausalMask (the skip fills mask_value "
-        "for all kv > q sub-tiles, assuming the mask masks exactly those); got "
-        f"{type(mask).__name__}. Disable qk_diag_skip for non-causal masks."
-    )
+  if config.qk_diag_skip:
```
Comment (octatrifan) [inline tokamax/_src/ops/experimental/tpu/splash_attention/splash_attention_kernel.py:2189]: nit: Mentioning the `qk_diag_skip` guards in the PR description (alongside a short description) would be beneficial

Comment (huytransformer) [reply tokamax/_src/ops/experimental/tpu/splash_attention/splash_attention_kernel.py:2189]: done

File: tokamax/_src/ops/experimental/tpu/splash_attention/splash_attention_mask_info.py (hunk 1 above)
Comment (huytransformer) [reply tokamax/_src/ops/experimental/tpu/splash_attention/splash_attention_mask_info.py:485]: Yep, and I can do a follow up for this one.

**Review by octatrifan — APPROVED** [review]
LGTM! Some interesting optimization ideas can arise from this.
